python-package/lightgbm/sklearn.py

```diff
@@ -11,14 +11,42 @@
 from .compat import (SKLEARN_INSTALLED, LGBMNotFittedError, _LGBMAssertAllFinite, _LGBMCheckArray,
                      _LGBMCheckClassificationTargets, _LGBMCheckSampleWeight, _LGBMCheckXY, _LGBMClassifierBase,
                      _LGBMComputeSampleWeight, _LGBMLabelEncoder, _LGBMModelBase, _LGBMRegressorBase, dt_DataTable,
-                     pd_DataFrame)
+                     pd_DataFrame, pd_Series)
 from .engine import train
 
+_ArrayLike = Union[List, np.ndarray, pd_Series]
+_EvalResultType = Tuple[str, float, bool]
+
+_LGBM_ScikitCustomObjectiveFunction = Union[
+    Callable[
+        [_ArrayLike, np.ndarray],
+        Tuple[_ArrayLike, _ArrayLike]
+    ],
+    Callable[
+        [_ArrayLike, np.ndarray, np.ndarray],
+        Tuple[_ArrayLike, _ArrayLike]
+    ],
+]
+_LGBM_ScikitCustomEvalFunction = Union[
+    Callable[
+        [_ArrayLike, np.ndarray],
+        Union[_EvalResultType, List[_EvalResultType]]
+    ],
+    Callable[
+        [_ArrayLike, np.ndarray, np.ndarray],
+        Union[_EvalResultType, List[_EvalResultType]]
+    ],
+    Callable[
+        [_ArrayLike, np.ndarray, np.ndarray, np.ndarray],
+        Union[_EvalResultType, List[_EvalResultType]]
+    ],
+]
```

Comment:
`y_true` cannot be anything else than `np.ndarray` because it is result of `get_label()` method:
https://github.com/microsoft/LightGBM/blob/144036f2df5b584b8302742c7036545f90f08fdb/python-package/lightgbm/sklearn.py#L173
https://github.com/microsoft/LightGBM/blob/144036f2df5b584b8302742c7036545f90f08fdb/python-package/lightgbm/basic.py#L2303-L2310
```suggestion
_LGBM_ScikitCustomObjectiveFunction = Union[
    Callable[
        [np.ndarray, np.ndarray],
        Tuple[_ArrayLike, _ArrayLike]
    ],
    Callable[
        [np.ndarray, np.ndarray, np.ndarray],
        Tuple[_ArrayLike, _ArrayLike]
    ],
]
_LGBM_ScikitCustomEvalFunction = Union[
    Callable[
        [np.ndarray, np.ndarray],
        Union[_EvalResultType, List[_EvalResultType]]
    ],
    Callable[
        [np.ndarray, np.ndarray, np.ndarray],
        Union[_EvalResultType, List[_EvalResultType]]
    ],
    Callable[
        [np.ndarray, np.ndarray, np.ndarray, np.ndarray],
        Union[_EvalResultType, List[_EvalResultType]]
    ],
]
```

Reply:
ah ok, thank you